Npm2100 driver

nordic-auko opened the pull request:
Add drivers for nPM2100 PMIC.
This is a multi-function device with regulators, gpio, watchdog, adc, and other supporting functions.

nordic-auko committed:
drivers: mfd: npm2100: Add npm2100 mfd driver

Add mfd driver for the npm2100 pmic device.
The driver contains basic initialization routines,
and functionality not covered by other device driver APIs.

Signed-off-by: Audun Korneliussen <audun.korneliussen@nordicsemi.no>

nordic-auko committed:
drivers: gpio: npm2100: Add driver for npm2100 pmic

Add gpio driver for npm2100 pmic.

Signed-off-by: Audun Korneliussen <audun.korneliussen@nordicsemi.no>

nordic-auko committed:
drivers: regulator: npm2100: Add driver for npm2100 pmic

Add regulator driver for npm2100 pmic.
This pmic has one boost and one ldo regulator.

Signed-off-by: Audun Korneliussen <audun.korneliussen@nordicsemi.no>

nordic-auko committed:
drivers: sensor: npm2100: Add driver for npm2100 pmic

Add sensor driver for npm2100 pmic.

This pmic performs measurements of battery voltage, regulator voltage
and die temperature.

Configurable pmic attributes are also organized under this driver.

Signed-off-by: Audun Korneliussen <audun.korneliussen@nordicsemi.no>

nordic-auko committed:
drivers: watchdog: npm2100: Add driver for npm2100 pmic

Add watchdog driver for npm2100 pmic.

This pmic has one timer that can be used for multiple functions,
including a watchdog that can reset or power-cycle connected devices.

Signed-off-by: Audun Korneliussen <audun.korneliussen@nordicsemi.no>

martinjaeger commented:
@nordic-auko this needs a rebase.

Your commits state "Initial driver for...". What does the "initial" mean? Is it still a draft or incomplete implementation? If not, I'd suggest to remove it to avoid any confusion.

nordic-auko commented:
> @nordic-auko this needs a rebase.
> 
> Your commits state "Initial driver for...". What does the "initial" mean? Is it still a draft or incomplete implementation? If not, I'd suggest to remove it to avoid any confusion.

Rebased it now.
The commit messages were based on the npm1300 pmic driver. I've changed the wording now to be more concise: it's not a draft or incomplete driver.

martinjaeger commented on dts/bindings/watchdog/nordic,npm2100-wdt.yaml:
```diff
@@ -0,0 +1,6 @@
+
+description: nPM2100 Watchdog
+
+compatible: "nordic,npm2100-wdt"
```
AFAIK `base.yaml` should be included for all drivers.

```suggestion
compatible: "nordic,npm2100-wdt"

include: base.yaml
```

gmarull replied:
I guess only if they use any base.yaml property

martinjaeger replied:
Yeah, it uses at least the `reg` property from `base.yaml`.

nordic-auko replied:
It doesn't use any properties other than what is inherited from the mfd parent device. I don't know if that involves base.yaml.

martinjaeger replied:
You are right. Should be ok not to import base.yaml.

martinjaeger commented on drivers/watchdog/wdt_npm2100.c:
```diff
@@ -0,0 +1,146 @@
+	return i2c_reg_write_byte_dt(&config->i2c, TIMER_TASKS_KICK, 1U);
+}
+
+static const struct wdt_driver_api wdt_npm2100_api = {
```
This should follow the new approach introduced in #82300.
```suggestion
static DEVICE_API(wdt, wdt_npm2100_api) = {
```

martinjaeger reviewed (changes requested):
Only reviewed the watchdog part, where I got a minor change request.

gmarull commented on drivers/mfd/Kconfig.npm2100:
```diff
@@ -0,0 +1,17 @@
+	help
+	  Enable the Nordic nPM2100 PMIC multi-function device driver
+
+config MFD_NPM2100_LEVEL_IRQ
+	bool "nPM2100 PMIC IRQ level"
+	depends on MFD_NPM2100
+	help
+	  Use level triggering instead of edge triggering when pin interrupt is enabled.
+	  This can reduce power consumption on some platforms, at the expense of irq latency.
```
shouldn't this be in DT?

nordic-auko replied:
Yeah, I was thinking about that, but wasn't sure exactly how to frame it. I'll add it as an enum property or bool then. As far as I understand, the GPIO DT flags doesn't include the interrupt flags

gmarull commented on drivers/mfd/mfd_npm2100.c:
```diff
@@ -0,0 +1,425 @@
+
+	gpio_init_callback(&mfd_data->gpio_cb, gpio_callback, BIT(config->host_int_gpios.pin));
+
+	ret = gpio_add_callback(config->host_int_gpios.port, &mfd_data->gpio_cb);
```
nit: use `gpio_add_callback_dt`

gmarull commented on drivers/gpio/gpio_npm2100.c:
```diff
@@ -0,0 +1,194 @@
+	return gpio_npm2100_port_set_masked_raw(dev, pins, ~value);
+}
+
+static int gpio_npm2100_pin_interrupt_configure(const struct device *dev, gpio_pin_t pin,
+						enum gpio_int_mode mode, enum gpio_int_trig trig)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(pin);
+	ARG_UNUSED(mode);
+	ARG_UNUSED(trig);
+
+	return -ENOTSUP;
+}
```
can be deleted

gmarull commented on drivers/gpio/gpio_npm2100.c:
```diff
@@ -0,0 +1,194 @@
+{
+	const struct gpio_npm2100_config *config = dev->config;
+
+	if (!i2c_is_ready_dt(&config->i2c)) {
+		return -ENODEV;
+	}
```
question: shouldn't register access go through MFD, so calls are _atomic_?

nordic-auko replied:
I don't fully understand. Do you mean having a i2c wrapper in the mfd driver, and protect that with a mutex?

gmarull commented on drivers/sensor/nordic/npm2100_vbat/npm2100_vbat.c:
```diff
@@ -0,0 +1,563 @@
+	}
+
+	/* Delay of vbat ADC measurement */
+	if ((chan == SENSOR_CHAN_GAUGE_VOLTAGE) &&
```
wondering if this should be a fuel gauge driver?

nordic-auko replied:
The npm2100 hardware only measures voltages, which I think fits better as a sensor device driver. To provide more full fledged fuel gauge functionality, additional software is required. We do have this kind of software in NCS/sdk-nrfxlib, and we are considering supporting the zephyr fuel gauge API, but I think that is a larger task outside the scope of this driver

gmarull reviewed (commented):
good stuff, a few comments from my side.

martinjaeger reviewed (dismissed):
Watchdog part OK.
